### PR TITLE
Include Translation methods to the DateTime and DateTimeImmutable

### PR DESCRIPTION
**Pull request by bordoni**
_Ref:_ [C#138196](http://central.tri.be/issues/138196)
---
📹 https://i.bordoni.me/YEuddmrP

Currently, we use the `DateTime` and `DateTimeImmutable` on our class handling when using `Date_Utils::build_date_object()`, which creates problems when translations are needed on the other end, because the code looks messy since you need to pass all strings that need to be translated to the `date_i18n` or `wp_date` functions.

The goal of this is to include two classes that will extend the `DateTime` and `DateTimeImmutable` so we can have a method for the translation to happen.

Example of Usage:
```php
Dates::build_date_object( $selected_start_date_value )->format_i18n( $date_formats->compact );
```

### CODE REVIEW

File: src/Tribe/Utils/Date_I18n.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	 *
+	 * @since  TBD
+	 *
+	 * @param  string $string Format to be used in the translation.
+	 *
+	 * @return string         Translated date.
+	 */
+	public function format_i18n( $string ) {
+		$formatted = $this->format( $string );
+		$unix_date =$this->format( 'U' );
```
Comment (sc0ttkclark) [inline src/Tribe/Utils/Date_I18n.php]:
```suggestion
		$unix_date = $this->format( 'U' );
```

File: src/Tribe/Utils/Date_I18n_Immutable.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	 *
+	 * @since  TBD
+	 *
+	 * @param  string $string Format to be used in the translation.
+	 *
+	 * @return string         Translated date.
+	 */
+	public function format_i18n( $string ) {
+		$formatted = $this->format( $string );
+		$unix_date =$this->format( 'U' );
```
Comment (sc0ttkclark) [inline src/Tribe/Utils/Date_I18n_Immutable.php]:
```suggestion
		$unix_date = $this->format( 'U' );
```

File: src/Tribe/Date_Utils.php [hunk 3]
```diff
@@ -1386,7 +1389,7 @@ static function immutable( $datetime = 'now', $timezone = null, $with_fallback =
 			}
 
 			if ( $datetime instanceof DateTime ) {
-				return DateTimeImmutable::createFromMutable( $datetime );
+				return ( new Date_I18n_Immutable() )->setTimestamp( DateTimeImmutable::createFromMutable( $datetime )->getTimestamp() );
```
Comment (lucatume) [inline src/Tribe/Date_Utils.php]: The conversion to timestamp is one that will drop the timezone information defaulting on the use of the system one.
I would rather implement a `Date_I18n_Immutable::createFromMutable` method (in `camelCase` to override the default PHP method and keep the API consistent) where we're never conflating the time information and the timezone information into a timestamp.

File: src/Tribe/Date_Utils.php [hunk 4]
```diff
@@ -1402,7 +1405,7 @@ static function immutable( $datetime = 'now', $timezone = null, $with_fallback =
 				return $cached;
 			}
 
-			$immutable = DateTimeImmutable::createFromMutable( $mutable );
+			$immutable = ( new Date_I18n_Immutable() )->setTimestamp( $mutable->getTimestamp() );
```
Comment (lucatume) [inline src/Tribe/Date_Utils.php]: The same comment about `createFromMutable` applies here.

File: src/Tribe/Date_Utils.php (hunk 3 above)
Comment (bordoni) [reply src/Tribe/Date_Utils.php]: Nice I like that.

**Comment by bordoni** [comment]
Timezone failing due to the time in which this ran.

File: src/Tribe/Utils/Date_I18n.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n.php]: file header is missing

File: src/Tribe/Utils/Date_I18n.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+namespace Tribe\Utils;
+
+use Tribe__Date_Utils as Dates;
+use DateTime;
+use DateTimeImmutable;
+
+class Date_I18n extends DateTime {
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n.php]: class header is missing

File: src/Tribe/Utils/Date_I18n_Immutable.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n_Immutable.php]: File header is missing

File: src/Tribe/Utils/Date_I18n_Immutable.php [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+namespace Tribe\Utils;
+
+use Tribe__Date_Utils as Dates;
+use DateTime;
+use DateTimeImmutable;
+
+class Date_I18n_Immutable extends DateTimeImmutable {
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n_Immutable.php]: Class header is missing

File: src/Tribe/Utils/Date_I18n_Immutable.php [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+	/**
+	 * Returns a translated string usign the params from this Immutable DateTime instance.
+	 *
+	 * @since  TBD
+	 *
+	 * @param  string $string Format to be used in the translation.
+	 *
+	 * @return string         Translated date.
+	 */
+	public function format_i18n( $string ) {
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n_Immutable.php]: Rename `$string` to `format`.

File: src/Tribe/Utils/Date_I18n.php [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	/**
+	 * Returns a translated string usign the params from this DateTime instance.
+	 *
+	 * @since  TBD
+	 *
+	 * @param  string $string Format to be used in the translation.
+	 *
+	 * @return string         Translated date.
+	 */
+	public function format_i18n( $string ) {
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n.php]: Rename `$string` to `$format`.

Comment (lucatume) [reply src/Tribe/Utils/Date_I18n.php]: Even better, to stick w/ what core uses, `$dateformatstring` or `$date_format`.

File: src/Tribe/Utils/Date_I18n.php [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+		$unix_date  = $this->format( 'U' );
```
Comment (lucatume) [inline src/Tribe/Utils/Date_I18n.php]: Codex says the `$unix_date` is "Unix timestamp, offset by the desired timezone."
Is the timezone offset taken into account here?

Comment (bordoni) [reply src/Tribe/Utils/Date_I18n.php]: Yeah it takes in cosideration.

**Pull request merged**